Example/street canyons

filipklg opened the pull request:
Creation of a new example for street canyon simulations.

Comparison between a simple diffuse case and BRDF retroreflection data for the simulation run.
A guide for using sparrowpy and visualization of BRDF input data.

Remarks:
The current implementation uses the same BRDF for the floor and the walls.
The current patch size is too big -> potentially poor final ETC but set low for the error finding.
Also different sized boundaries compared to Kang to speed up the simulation (paper boundary size example in code comment).
Using the same source and receiver position to demonstrate the effect of retroreflection.

Problems:
Without the usage of numba: "Casting complex values to real discards the imaginary part" warning in _form_factors_with_directivity_dim when "scattering is not None".
Numba runs may encounter "Pass nopython_type_inference" at the same code location.
--- Provisional data check doesn't reproduce the cause of the "imaginary value" warning for the BRDF data(!?)

Additionally:
Small typo FIX in the existing file example\fast_radiosity

filipklg committed:
Close to final but still bake_geometry faulty. Additionally, typo in fast_radiosity example fixed

filipklg committed:
New vers

filipklg committed:
FIX: partly BRDF setup

False import of data and slight changes to 2 freq bands
Still NUMBA ERROR (without fine)

filipklg committed:
REFACTOR: Divide parts and plot

better visuals
Plot v1

filipklg committed:
ADD: simple plotting

scatter plotting (prob faulty)
still issues with real/imag bakegeometry

filipklg committed:
FIX: plotting spharpy

result more useful but return angle max?

filipklg committed:
ADD: imag value check

no imag values contained in the brdf data

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "brdf_receivers = pf.Coordinates(0, 0, 1, weights=1)\n",
+    "\n",
+    "# set the coefficient values per frequency band\n",
+    "frequencies = np.array([2000, 4000])\n",
```
```suggestion
    "frequencies = np.array([2000])\n",
```

one frequency band is sufficiant for demonstration

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    ")\n",
+    "\n",
+    "# short data check\n",
+    "print(\"\\nThe BRDF data contains\", brdf_sources_rr.csize, \"points.\")\n",
```
we can replace this cell by a generic BRDF representing the retro-reflection. e.g. 50% retroreflection
```python
brdf_in = pf.samplings.sph_gaussian(sh_order=15)
brdf_in = brdf_in[brdf_in.z > 0]
brdf_out = brdf_in.copy()

bsc_retro_reflection = np.ones((brdf_in.csize, brdf_out.csize, len(frequencies)))
bsc_retro_reflection /= (2*brdf_out.csize)
bsc_retro_reflection[np.arange(brdf_in.csize), np.arange(brdf_out.csize), 0] += .5
bsc_retro_reflection = pf.FrequencyData(bsc_retro_reflection, frequencies)

brdf_retro = sp.brdf.create_from_directional_scattering(
    brdf_in, brdf_out,
    bsc_retro_reflection,
    absorption_coefficient=absorption_data,
    )


radiosity_retroreflection.set_wall_brdf(
    np.arange(len(boundaries)).tolist(),
    brdf_retro,
    brdf_in,
    brdf_out,
)
```

for me this was working. also with numba

joao-fatela replied:
I agree that this is the more elegant solution. This way we can also control the number of BRDF samples in the simulation, which we can optimize for a bit better performance.

ahms5 replied:
the example I gave was not lambertian, due to the non constant energy in the brdf, we can move to sth like:

```
brdf_in = pf.samplings.sph_gaussian(sh_order=15)
brdf_in = brdf_in[brdf_in.z > 0]
brdf_out = brdf_in.copy()

bsc_retro_reflection = np.zeros((brdf_in.csize, brdf_out.csize, len(frequencies)))
bsc_retro_reflection[np.arange(brdf_in.csize), np.arange(brdf_out.csize), 0] += .5
bsc_retro_reflection = pf.FrequencyData(bsc_retro_reflection, frequencies)

brdf_retro = sp.brdf.create_from_directional_scattering(
    brdf_in, brdf_out,
    bsc_retro_reflection,
    absorption_coefficient=absorption_data,
    )
brdf_retro.freq += 1/2/np.pi


radiosity_retroreflection.set_wall_brdf(
    np.arange(len(boundaries)).tolist(),
    brdf_retro,
    brdf_in,
    brdf_out,
)

```

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "This example shows the process of simulating a geometry such as a street canyon while applying different properties to the wall boundaries. The setup of the canyon is comparable to **Sound propagation in street canyons: Comparison between\n",
+    "diffusely and geometrically reflecting boundaries** by Kang published in 2000.\n",
+    "\n",
+    "Paper: https://doi.org/10.1121/1.428580"
```
maybe we need to remove the reference, since with the retroreflection is has nothing to do with the origial reference anymore.

joao-fatela replied:
I second this

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "patch_size = 4  # small values increase computation time by a lot\n",
+    "etc_duration = 1\n",
+    "etc_time_resolution = 1 / 1000\n",
+    "max_reflection_order = 90\n",
```
```suggestion
    "etc_time_resolution = 1 / 100\n",
    "max_reflection_order = 50\n",
```
I would reduce it, otherwise the example will take too much time

joao-fatela replied:
the reflection order can go even lower, depending on the final geometry used in this example. It will have to be tuned when we finalize this example

ahms5 reviewed (changes requested):
Thank you Filip, I would simplify it a bit, see comments, otherwise it looks fine.

filipklg commented:
Note: When changing to only one frequency band in the simple case -> remove plotting of two freq. bands for the first simple instance at the end!

ahms5 commented:
> Note: When changing to only one frequency band in the simple case -> remove plotting of two freq. bands for the first simple instance at the end!

Yes, thats fine, I would just plot the two etcs of the differnt method into one plot to keep the example simple.

onces you incled the the suggestions or commented on them if you disagree, please rerequest the reiew.

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "frequencies = np.array([2000, 4000])\n",
+    "scattering_coefficient = 1  # totally diffuse\n",
+    "scattering_data = pf.FrequencyData(\n",
+    "    [scattering_coefficient, scattering_coefficient],\n",
```
```suggestion
    "    scattering_coefficient,\n",
```
Like Anne mentions in the previous comment comment, it's best to use a single frequency for demonstration purposes. Idem for the absorption coefficient below.

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define the street canyon and set general parameters for the simulation."
```
maybe split the two?
e.g.:

"Definition of the street canyon"
-- geometry definition code --
"general simulation parameter definition"
-- simulation parameter definition --

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "# canyon boundary dimensions in meters 120 20 18\n",
+    "length = 35  # X\n",
+    "width = 10  # Y\n",
+    "height = 4  # Z  (different heights were used by Kang)\n",
```
not relevant if Kang citation is removed

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
```
before this cell I think it could be good to show an illustration (so axes x, y, and z make sense to the user)

maybe a markdown cell describing the scene

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "    brdf_receivers_rr.z[direction],\n",
+    ")\n",
+    "ind_max_reflection = np.argmax(brdf_rr.freq[direction, :, -1])\n",
+    "print(\n",
```
Not sure if this is important. don't forget to remove stuff you use for debugging

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To run the simulation, the geometry must first be calculated (baked) to enable the computation of energy exchange between patches at a later stage.\n",
```
```suggestion
    "To run the simulation, the energetic relationships of the different surfaces of the geometry are precomputed (baked) to enable the energy exchange estimation at a later stage.\n",
```

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "source": [
+    "To run the simulation, the geometry must first be calculated (baked) to enable the computation of energy exchange between patches at a later stage.\n",
+    "\n",
+    "The patch to patch relationships contain the form factors, the geometrical relationship between the patches as well as the BRDFs.\n",
```
```suggestion
```

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "\n",
+    "The patch to patch relationships contain the form factors, the geometrical relationship between the patches as well as the BRDFs.\n",
+    "\n",
+    "This is the most heavy part of the simulation and is independent of the source and receiver positions."
```
```suggestion
    "This part of the simulation is independent of the source and receiver positions."
```

joao-fatela commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Subsequently, the energy exchange between the patches and the energy at the receiver is calculated."
```
```suggestion
    "Subsequently, the energy exchange between the patches is calculated."
```

joao-fatela reviewed (changes requested):
Good job, thank you so much.
In general, I think it can still get a bit more user friendly: The language can be more succint, for example.

I agree with Anne's comments, there's a few things that can be shaved in more elegant solutions.

Remember to re-request the revision once you've made the suggested changes (or argued for your solutions)

filipklg committed:
Style&Simplified

Better cell structure and clearer Markdown texts
Only one frequency band for the simple simulation

filipklg committed:
Style

Two text changes

filipklg committed:
FEATURE

Plot for geometry

filipklg committed:
CHANGE: BRDF input

Generic input and different text

filipklg committed:
FIX: Debug and Speed

Cleanup of test parts
Shorter runtime

filipklg commented:
Pull request only adds this Notebook (and a small textchange to another Notebook).
PROBLEM:
Why is suddenly pytest_blender failing?

Changes:
Adopted the proposed way of using a generic BRDF for showing the effect of retroreflection instead of data from a file.
Some text blocks were streamlined.
Selection of more or less optimal values for speed of calc and quality of results.
Still a reference to the streetcanyon publication but better integrated imo (gives context).
Visualization of the geometry (and the created BRDF positions).

Potentially missing:
Further explanation of the process of defining the (currently) 50% retroreflection for the same indices of identical incoming and outgoing brdf datapoint angles - maybe already clear enough?

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -50,19 +45,55 @@
+    "\n",
+    "\n",
+    "Dimensions for boundaries can be set as in **Sound propagation in street canyons: Comparison between\n",
+    "diffusely and geometrically reflecting boundaries** by Kang published in 2000. Paper: https://doi.org/10.1121/1.428580"
```
please remove kang reference, it is not relevant hree at all.

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -50,19 +45,55 @@
+   ],
+   "source": [
+    "# Canyon boundary dimensions in meters\n",
+    "# (Kang used e.g. 120x20x18 but this takes longer to compute)\n",
```
```suggestion
```

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "            ],\n",
+    "        ),\n",
+    "        np.array([1, 0, 0]),  # up vector\n",
+    "        np.array([0, 1, 0]),  # normal vector\n",
```
```suggestion
    "        up_vector=np.array([1, 0, 0]),\n",
    "        normal=np.array([0, 1, 0]),\n",
```

this would also work

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "    np.array([[width, width], [width, width]]),\n",
+    "    np.array([[0, 0], [height, height]]),\n",
+    "    alpha=0.8,\n",
+    ")\n",
```
```suggestion
for boundary in boundaries:
    ax.plot_surface(
        boundary.pts[:, 0].reshape((2, 2)),
        boundary.pts[:, 1].reshape((2, 2)),
        boundary.pts[:, 2].reshape((2, 2)),
        alpha=0.8,
    )
```

the plot shuld be based on the boundaries. later on (after this pr is merged) this can be replaces by using the new method #140

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "    alpha=0.8,\n",
+    ")\n",
+    "ax.set_box_aspect((3.5, 1, 1))\n",
+    "ax.set_zlim(0, 10)\n",
```
```suggestion
    "ax.set_aspect('equal')\n",
```

aspect ratio should be eqaul if all axis have the same unit

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "\n",
+    "# Plotting the scene\n",
+    "fig = plt.figure()\n",
+    "ax = fig.add_subplot(projection=\"3d\")\n",
```
we could also add source and receiver positons in the plot.
```python
ax.scatter(source.x, source.y, source.z, color="red", label="Source")
ax.scatter(receiver.x, receiver.y, receiver.z, color="blue", label="Receiver")
```

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "    brdf_out.z,\n",
+    ")\n",
+    "plt.title(\"BRDF sampling points\")\n",
+    "plt.show()"
```
not sure it we really need it, if you want to keep it, use [pyfar.plot.scatter](https://pyfar--872.org.readthedocs.build/en/872/modules/pyfar.plot.html#pyfar.plot.scatter) instead

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,509 @@
+    "    linestyle=\"--\",\n",
+    ")\n",
+    "pf.plot.time(\n",
+    "    etc_retroreflection[0, -1],  # plot only the highest frequency band (-1)\n",
```
```suggestion
    "    etc_retroreflection,\n",
```
not needed anymore

ahms5 reviewed (changes requested):
thank you Filip, looks nice overall, I hae just some suggestions for the plotting

ahms5 commented:
the tests are fixed on develop, please merge develop into your branch.

filipklg committed:
CHANGE and FIX:
Fully remove Kang reference
Remove one plot
VIS

filipklg committed:
Merge remote-tracking branch 'origin/develop' into example/street_canyons

filipklg committed:
FIX FILE PYPROJECT
Add "spharpy" to the list for the tests running in github for the merge

ahms5 reviewed (approved):
thanks filip, good to merge for me, @joao-fatela, what do you think?

filipklg commented:
Just a quick note:
I have added spharpy in the pyproject file so github pytest is completing and someone installing sp can easily use the exampe right?

joao-fatela reviewed (commented):
Great job, I think it can be merged.

_If you guys agree_, I think it would be helpful to plot the different BRDF, so that it is a bit more intuitive for users to understand how they differ (just reading the code isn't particularly illuminating). 

In any case, I don't think that this is a major reason not to merge. Just let me know whether you think this is useful @ahms5 @filipklg and if you disagree I'll approve the merge.

filipklg commented:
Yeah give me one second, I can create a new commit with a comparison of the hemispheres

filipklg committed:
ADD plotting BRDF

filipklg commented:
If you approve of the visualization, just merge it. Otherwise, feel free to change it or remove it (as in the previous commit). Unfortunately, I get deprecation warnings that I can't easily remove (it doesn't look clean to me).

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    }
+   ],
+   "source": [
+    "# Plot the simple BRDF with one value equal to every direction\n",
```
I would suggest to use the [baloon plot](https://spharpy.readthedocs.io/en/stable/modules/spharpy.plot.html#spharpy.plot.balloon) instead, we also need to show the incident direction, maybe by a line, otherwise it is not really telling anything.

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "source": [
+    "# Plot the simple BRDF with one value equal to every direction\n",
+    "fig = plt.figure()\n",
+    "ax = fig.add_subplot(projection=\"3d\")\n",
```
both axes could be in one figure, see https://matplotlib.org/stable/gallery/subplots_axes_and_figures/subplots_demo.html

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+    "ax = fig.add_subplot(projection=\"3d\")\n",
+    "sph.plot.pcolor_sphere(\n",
+    "    brdf_out,\n",
+    "    np.repeat(brdf_diffuse.freq, brdf_out.csize),\n",
```
this looks very hacky, how about?
```suggestion
    "    brdf_diffuse.freq[direction, :, -1],\,\n",
```

ahms5 commented on pyproject.toml:
```diff
@@ -55,6 +55,7 @@ tests = [
     "ipykernel",
     "sparrowpy[fast]",
     "nbmake",
+    "spharpy",
 ]
```
```suggestion
]
examples = [
    "spharpy",
]
```

I woudl suggest to have a dedicated option for examples. ``examples`` need to be added to dev as well

```
dev = ["sparrowpy[deploy,tests,docs,blender,examples]"]
```

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "source": [
+    "import numpy as np\n",
+    "import pyfar as pf\n",
+    "import spharpy as sph\n",
```
```suggestion
    "import spharpy\n",
```

please import as spharpy.

ahms5 commented on examples/street_canyon_radiosity.ipynb:
```diff
@@ -0,0 +1,523 @@
+   "metadata": {},
+   "source": [
+    "\n",
+    "First, import all of our dependencies."
```
add a note that spharpy need to be installed

ahms5 reviewed (changes requested):
Thank you, there are now more thinks to consider.

filipklg committed:
CHANGE: pyproject+plotting

ballon okay?
improved incident angle

ahms5 commented:
We decided to remove the plot and remove the spharpy dependency here.
